### ISSUE TITLE
Demo track 9 - Frequency cutoff at 11 kHz

### ISSUE THREAD

**Issue by JeffreyCA** [ws-choi/Conditioned-Source-Separation-LaSAFT#3]
This looks promising, very nice work!
It looks like the separated parts for the **footprint** track cut off at 11 kHz, [similar to Spleeter](https://github.com/deezer/spleeter/wiki/5.-FAQ#why-are-there-no-high-frequencies-in-the-generated-output-files-). Do you know what's going on here? The other tracks don't have this cut off.

<img width="1184" alt="Screen Shot 2020-12-21 at 9 39 27 PM" src="https://user-images.githubusercontent.com/9157833/102842478-0bccd380-43d5-11eb-97b7-05a9101f63e4.png">

**Comment by ws-choi** [comment]
Hi JeffreyCA,
Is this image the output of our model?
Can you share the separated outputs in the wav or mp3 format?

- To clarify, 
  - our model takes an audio file (44100hz)
  - it applies stft (n_fft:2048 or 4096) on the audio file to obtain the complex-valued spectrgram
  - it does not cut off any freq
  - it estimates the target complex-valued spectrogram of the target
  - it reconstructs signal by applying i-stft.

TLDR; our models use the full coverage of the frequency-axis of the given spectrogram.

**Comment by JeffreyCA** [comment]
Hi, I was looking at the output files on your demo page. The outputs for Track 9 (footprint) are cut off at 11k but the other songs are not.

https://lasaft.github.io/audios/footprint.mp3
https://lasaft.github.io/audios/footprint-vocals.wav
https://lasaft.github.io/audios/footprint-bass.wav
https://lasaft.github.io/audios/footprint-drums.wav
https://lasaft.github.io/audios/footprint-other.wav

How come it's only this track that is cut off?

**Comment by JeffreyCA** [comment]
Btw I also noticed on the demo page under Track Infomation, track 8 is "Footprints - Woosung Choi" but in the table track 9 is "Footprints".

**Comment by ws-choi** [comment]
Thank you, I'll revise it later 👍 
It is very interesting result since I used the same scripts for separating sources of footprint.

I guess that this is because footprint is the original track whose format is 'mp3' with sample rate of 22050.
The other tracks are wav files
![화면 캡처 2020-12-22 122027](https://user-images.githubusercontent.com/13086447/102845284-90106c80-4450-11eb-9dfa-4b4f35ccb656.png)
.

**Comment by JeffreyCA** [comment]
I checked the sample rate of `footprint.mp3` on my local machine and it says it's 44.1 kHz. I think when you do `librosa.load` without specifying the `sr` parameter it defaults to 22150, see [this](https://librosa.org/blog/2019/07/17/resample-on-load/).

**Comment by ws-choi** [comment]
Yes I was wrong. I also checked it.
Below is the script for separating them.
I put all tracks in the ```demo``` directory and ran the script below:

```python
import os
import librosa
import soundfile

directory = 'demo'
for filename in os.listdir(directory):
    if filename.endswith(".wav") or filename.endswith(".mp3"):
        path = os.path.join(directory, filename)
        print(path)
        data, sr=librosa.load(path, mono=False)
        data = librosa.resample(data, sr, 44100)
        
        for target in ['vocals', 'drums','bass','other']:
          model.separate_track(data.T, target)
          os.rename('temp.wav',os.path.join('result', filename[:-4]+'_'+ target+'.wav')) 

    else:
        continue
```

I'll find out what have happened sooner or later

**Comment by ws-choi** [comment]
bty, how did you create the spectrogram image above??

**Comment by JeffreyCA** [comment]
> bty, how did you create the spectrogram image above??

I used Audacity ([reference](https://manual.audacityteam.org/man/spectrogram_view.html#Context_menu))

**Comment by JeffreyCA** [comment]
I just ran "footprints" through your colab notebook and the higher frequencies are all there.

**Comment by ws-choi** [comment]
> I just ran "footprints" through your colab notebook and the higher frequencies are all there.

Did you run the script for the original only, or both the original and all the separated files? And can you share the script you with me used?

**Comment by JeffreyCA** [comment]
I did not run the script above. I ran the following in your colab notebook:
```
!wget https://lasaft.github.io/audios/footprint.mp3
# ...
# load model, etc
# ...
audio, rate = librosa.load('footprint.mp3', sr=44100, mono=False)
separated = model.separate_track(audio.T, 'drums')
```

Then I downloaded the `temp.wav` to my computer and checked the spectrogram.

**Comment by ws-choi** [comment]
I think I finally got a clue!

For short-duration files (<15 secs), it seems that the spectrogram of the separated file has high freqs when you use **AUDACITY** for spectrogram analysis.

I installed audacity and used spec view with max freq: 22050 and window size: 1024-(default).

<img width="940" alt="spec" src="https://user-images.githubusercontent.com/13086447/102952978-aa5f4e80-4513-11eb-959f-de0565740c6a.png">

Try this file: https://github.com/lasaft/lasaft.github.io/blob/master/audios/shortprint.wav
The file is a short version of the 'footprint' track.

---

I guess this issue was caused by very complex reasons such as  AUDACITY's visualization methods.

BTW, our [models](https://github.com/ws-choi/Conditioned-Source-Separation-LaSAFT/blob/0a98ee9d5a0220d5a61c3d76fd60d45b586e7e31/lasaft/source_separation/conditioned/cunet/dcun_base.py#L189) do not exploit an explicit 'Frequency cutoff'.
Regardless of the length of given input audio, it applies the same [script](https://github.com/ws-choi/Conditioned-Source-Separation-LaSAFT/blob/0a98ee9d5a0220d5a61c3d76fd60d45b586e7e31/lasaft/source_separation/conditioned/cunet/dcun_base.py#L250) for separation.

---

TLDR; for short duration files (<15 secs), it seems that the spectrogram of the separated file has high freqs when you use **AUDACITY** for spectrogram analysis. Our [models](https://github.com/ws-choi/Conditioned-Source-Separation-LaSAFT/blob/0a98ee9d5a0220d5a61c3d76fd60d45b586e7e31/lasaft/source_separation/conditioned/cunet/dcun_base.py#L189) do not exploit an explicit 'Frequency cutoff'.

**Comment by JeffreyCA** [comment]
Sorry, but I'm not sure I understand...

To make it more clear, I uploaded two files here: https://github.com/JeffreyCA/footprint/

**JeffreyCA-footprint-vocals.wav** is what I generated through your collab to isolate the vocals of `footprint.mp3`.
**original-footprint-vocals.wav** is the same file as https://lasaft.github.io/audios/footprint-vocals.wav.

If you use Audacity or any other spectrogram tool, they are clearly different:
<img width="1410" alt="Screen Shot 2020-12-22 at 11 05 27 PM" src="https://user-images.githubusercontent.com/9157833/102958137-3f2d6200-44aa-11eb-86de-476a2dc5bb94.png">

In **original-footprint-vocals.wav**, there's nothing above 11 kHz (which shouldn't be happening), but in `JeffreyCA-footprint-vocals.wav` those frequencies are there (this is expected).

**JeffreyCA-footprint-vocals.wav**:
<img width="514" alt="Screen Shot 2020-12-22 at 11 07 52 PM" src="https://user-images.githubusercontent.com/9157833/102958335-9f240880-44aa-11eb-977b-7baf96badf2b.png">

**original-footprint-vocals.wav**:
<img width="504" alt="Screen Shot 2020-12-22 at 11 08 15 PM" src="https://user-images.githubusercontent.com/9157833/102958348-a519e980-44aa-11eb-931c-f68ddfe48b14.png">

**Comment by ws-choi** [comment]
Have you ever tried this?
<img width="177" alt="1" src="https://user-images.githubusercontent.com/13086447/102958896-58bdc080-4521-11eb-88ce-ef1a59668294.png">
<img width="323" alt="2" src="https://user-images.githubusercontent.com/13086447/102958901-5b201a80-4521-11eb-9b01-0c534d8c5197.png">

It seems AUDACITY automatically filters out relatively 'noisy' freqs, for better visualization.

**Comment by ws-choi** [comment]
<img width="928" alt="3" src="https://user-images.githubusercontent.com/13086447/102959119-00d38980-4522-11eb-99fd-9bfefcb04032.png">

spec analysis of separated 'vocals' file of 'shortprint.wav'

**Comment by JeffreyCA** [comment]
You're right, to clarify, all the other demo tracks **are fine**. It's just the **footprint** track that's weird.
Here's a different tool: https://academo.org/demos/spectrum-analyzer/. If you upload the footprint-vocals.wav and play it, you see no high frequencies. If you upload my version, they are there.

**Comment by ws-choi** [comment]
Oh, it's very weird. You are right. 
I'm very sorry for my misunderstanding. 
I'm re-generating files for footprints, and I'll re-post it when it finished. 
It seems these files have high frequencies.
Thank you very much!

**Comment by JeffreyCA** [comment]
Thank you, the new tracks look good!

Another thing, how come if you add up all the individual sources (vocals + other + bass + drums), it doesn't sound the same as the original? If you focus on the higher frequencies of the vocals or hi-hats they sound muffled compared to the original. It's not just this model, it seems like a common observation I've seen across other source separation models as well.

**Comment by ws-choi** [comment]
It's a good point. I recommend you to read the section [Energy Preserved Wasserstein Learning](https://arxiv.org/abs/1711.04121) of this paper [1].

As mentioned in the paper: 
> the loss function involves i) the energy preservation term to restrict the separated sources's total energy is close to the mixed one 

Trained with this auxiliary loss function, a source separation model can produce better results. The sum of its separated results would be closer to the original. 

We have not exploited the auxiliary loss function for training our models for the sake of simplicity. 

[1] Zhang, Ning, Junchi Yan, and Yuchen Zhou. "Weakly supervised audio source separation via spectrum energy preserved wasserstein learning." arXiv preprint arXiv:1711.04121 (2017).

**Comment by JeffreyCA** [comment]
Thanks for your help! I'll go ahead and close this issue now.
